kuksa_databroker/databroker/src/permissions.rs

```diff
@@ -13,56 +13,207 @@
+        match permission {
+            Permission::Nothing => {
+                // Adding nothing
+            }
+            Permission::All => self.provide.extend_with(PathMatchBuilder::Everything),
+            Permission::Glob(path) => self.provide.extend_with_glob(path),
+        };
+        self
+    }
+
+    pub fn add_create_permission(mut self, permission: Permission) -> Self {
+        match permission {
+            Permission::Nothing => {
+                // Adding nothing
+            }
+            Permission::All => self.create.extend_with(PathMatchBuilder::Everything),
+            Permission::Glob(path) => self.create.extend_with_glob(path),
+        };
+        self
+    }
+
+    pub fn build(self) -> Result<Permissions, PermissionsBuildError> {
+        Ok(Permissions {
+            expires_at: self.expiration,
+            read: self.read.build()?,
+            actuate: self.actuate.build()?,
+            provide: self.provide.build()?,
+            create: self.create.build()?,
+        })
+    }
+}
+
+impl Permissions {
+    pub fn builder() -> PermissionBuilder {
+        PermissionBuilder::new()
+    }
+
+    pub fn can_read(&self, path: &str) -> Result<(), PermissionError> {
+        self.expired()?;
+
         if self.read.is_match(path) {
             return Ok(());
         }
-        Err(Error::AccessDenied)
-    }
 
-    pub fn actuate_allowed(&self, path: &str) -> Result<(), Error> {
-        if self.expires_at < SystemTime::now() {
-            return Err(Error::Expired);
-        }
+        // Read permissions are included (by convention) in the
+        // other permissions as well.
         if self.actuate.is_match(path) {
             return Ok(());
         }
-        Err(Error::AccessDenied)
+        if self.provide.is_match(path) {
+            return Ok(());
+        }
+        if self.create.is_match(path) {
+            return Ok(());
+        }
+
+        Err(PermissionError::Denied)
     }
 
-    pub fn provide_allowed(&self, path: &str) -> Result<(), Error> {
-        if self.expires_at < SystemTime::now() {
-            return Err(Error::Expired);
+    pub fn can_write_actuator_target(&self, path: &str) -> Result<(), PermissionError> {
+        self.expired()?;
+
+        if self.actuate.is_match(path) {
+            return Ok(());
         }
+        Err(PermissionError::Denied)
+    }
+
+    pub fn can_write_datapoint(&self, path: &str) -> Result<(), PermissionError> {
+        self.expired()?;
+
         if self.provide.is_match(path) {
             return Ok(());
         }
-        Err(Error::AccessDenied)
+        Err(PermissionError::Denied)
+    }
+
+    pub fn can_create(&self, path: &str) -> Result<(), PermissionError> {
+        self.expired()?;
+
+        if self.create.is_match(path) {
+            return Ok(());
+        }
+        Err(PermissionError::Denied)
+    }
+
+    #[inline]
+    pub fn expired(&self) -> Result<(), PermissionError> {
```

Comment:
shouldn't this function be above the first use of it? For reading and understanding purposes.